src/prof_recent.c

```diff
@@ -0,0 +1,407 @@
+		assert(nstime_compare(&recent->dalloc_time, &zero_time) == 0);
+		nstime_update(&recent->dalloc_time);
+		assert(recent->dalloc_tctx == NULL);
+		recent->dalloc_tctx = dalloc_tctx;
+	} else {
+		/* We lost the rase - the allocation record was just gone. */
+		decrement_recent_count(tsd, dalloc_tctx);
+	}
+	malloc_mutex_unlock(tsd_tsdn(tsd), &prof_recent_alloc_mtx);
+}
+
+static void
+prof_recent_alloc_evict_extent(tsd_t *tsd, prof_recent_t *recent) {
+	malloc_mutex_assert_owner(tsd_tsdn(tsd), &prof_recent_alloc_mtx);
+	if (recent->alloc_extent != NULL) {
+		extent_prof_recent_alloc_reset(recent->alloc_extent, recent);
+	}
+}
+
+static void
+assert_prof_recent_alloc_non_empty(tsd_t *tsd) {
+	malloc_mutex_assert_owner(tsd_tsdn(tsd), &prof_recent_alloc_mtx);
+	assert(prof_recent_alloc_dummy != NULL);
+	assert(prof_recent_alloc_dummy->next != prof_recent_alloc_dummy);
+	assert(prof_recent_alloc_count > 0);
+}
+
+void
+prof_recent_alloc(tsd_t *tsd, prof_tctx_t *tctx, const void *ptr, size_t usize,
+    nstime_t t) {
+	extent_t *extent = iealloc(tsd_tsdn(tsd), ptr);
+	assert(extent != NULL);
+	assert(tctx != NULL && tctx == extent_prof_tctx_get(extent));
+
+	malloc_mutex_assert_not_owner(tsd_tsdn(tsd), tctx->tdata->lock);
+	malloc_mutex_lock(tsd_tsdn(tsd), &prof_recent_alloc_mtx);
+
+	if (prof_recent_alloc_max == 0) {
+		goto label_rollback;
+	}
+
+	if (prof_recent_alloc_dummy == NULL) {
+		/*
+		 * This is only encountered when the very first node is being
+		 * added - allocate the dummy node and circle it back.
+		 */
+		prof_recent_alloc_dummy = (prof_recent_t *)iallocztm(
+		    tsd_tsdn(tsd), sizeof(prof_recent_t),
+		    sz_size2index(sizeof(prof_recent_t)), false, NULL, true,
+		    arena_get(tsd_tsdn(tsd), 0, false), true);
+		if (prof_recent_alloc_dummy == NULL) {
+			goto label_rollback;
+		}
+		prof_recent_alloc_dummy->next = prof_recent_alloc_dummy;
+	}
+	assert(prof_recent_alloc_dummy != NULL);
+
+	{
+		/* Fill content into the dummy node. */
+		prof_recent_t *node = prof_recent_alloc_dummy;
+		node->usize = usize;
+		nstime_copy(&node->alloc_time, &t);
+		node->alloc_tctx = tctx;
+		extent_prof_recent_alloc_set_unchecked(extent, node);
+		nstime_copy(&node->dalloc_time, &zero_time);
+		node->dalloc_tctx = NULL;
+	}
+
+	prof_tctx_t *old_alloc_tctx, *old_dalloc_tctx;
+
+	if (prof_recent_alloc_count == prof_recent_alloc_max) {
+		/* If upper limit is reached, simply shift the dummy. */
+		assert(prof_recent_alloc_max != -1);
+		assert_prof_recent_alloc_non_empty(tsd);
+		prof_recent_alloc_dummy = prof_recent_alloc_dummy->next;
+		old_alloc_tctx = prof_recent_alloc_dummy->alloc_tctx;
+		old_dalloc_tctx = prof_recent_alloc_dummy->dalloc_tctx;
+		prof_recent_alloc_evict_extent(tsd, prof_recent_alloc_dummy);
+	} else {
+		/* Otherwise allocate a new node to serve as the dummy. */
+		assert(prof_recent_alloc_max == -1 ||
+		    prof_recent_alloc_count < prof_recent_alloc_max);
+		prof_recent_t *node = (prof_recent_t *)iallocztm(tsd_tsdn(tsd),
```

Comment:
If we have more than one place allocating the node, add a separate function. (I'm hoping we don't have to, if we statically define the dummy node then this would be the only place allocating)

Reply:
Yes, we have two places allocating the node, but the calls are not the same: in booting, we don't have a TSD, so I need to use `TSDN_NULL`, as well as `arena_get(TSDN_NULL, 0, true)`.